File: src/main/java/org/c3lang/intellij/psi/C3CallablePsiElement.java
package org.c3lang.intellij.psi;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import java.util.List;

public interface C3CallablePsiElement extends C3FullyQualifiedNamePsiElement
{
	@NotNull String getSourceFileName();
	@Nullable ShortType getType();
	@Nullable ShortType getReturnType();
	@NotNull List<ParamType> getParameterTypes();
}

File: src/main/java/org/c3lang/intellij/psi/C3CallablePsiElement.kt
package org.c3lang.intellij.psi

interface C3CallablePsiElement : C3FullyQualifiedNamePsiElement {
    val sourceFileName: String
    val type: ShortType?
    val returnType: ShortType?
    val parameterTypes: List<ParamType>
    val parameterListString: String get() = parameterTypes.joinToString(", ") {
        listOfNotNull(
            it.type?.fullName,
            it.name
        ).joinToString(" ")
    }
}